### PR TITLE
Upgrade macOS build to macOS 14

### PR DESCRIPTION
**Pull request by ann0see**
**Short description of changes**

* This will also build on arm runners
* Switches build directory to more standard location on Apple silicon macs
* Updates to use virtual environment for python as per https://docs.brew.sh/Python-for-Formula-Authors#python-declarations-for-applications

CHANGELOG: Upgrade main macOS build to macOS 14 and Apple Silicon build

**Context: Fixes an issue?**

Fixes: #3296 
**Does this change need documentation? What needs to be documented and how?**
No

**Status of this Pull Request**

Ready for testing

**What is missing until this pull request can be merged?**

Review and test
## Checklist

-  [x] I've verified that this Pull Request follows the [general code principles](https://github.com/jamulussoftware/jamulus/blob/main/CONTRIBUTING.md#jamulus-projectsource-code-general-principles)
-  [x] I tested my code and it does what I want
-  [x] My code follows the [style guide](https://github.com/jamulussoftware/jamulus/blob/main/CONTRIBUTING.md#source-code-consistency) 
-  [x] I waited some time after this Pull Request was opened and all GitHub checks completed without errors. 
-  [x] I've filled all the content above

### CODE REVIEW

File: .github/autobuild/mac.sh [hunk 1]
```diff
@@ -26,7 +26,7 @@
 
 set -eu
 
-QT_DIR=/usr/local/opt/qt
+QT_DIR=/opt/qt
```
Comment (ann0see) [inline .github/autobuild/mac.sh:29]: To fit to modern apple silicon maps similar to https://docs.brew.sh/Installation

File: .github/autobuild/mac.sh [hunk 2]
```diff
@@ -47,7 +47,15 @@ setup() {
         echo "Using Qt installation from previous run (actions/cache)"
     else
         echo "Installing Qt..."
-        python3 -m pip install "aqtinstall==${AQTINSTALL_VERSION}"
+        # We may need to create the Qt installation directory and chown it to the runner user to fix permissions
+        sudo mkdir "${QT_DIR}"
+        sudo chown "$(whoami)" "${QT_DIR}"
+        # Create and enter virtual environment
+        python3 -m venv venv
+        # Must hide directory as it just gets created during execution of the previous command and cannot be found by shellcheck
```
Comment (ann0see) [inline .github/autobuild/mac.sh:55]: Unfortunately needed.

**Comment by ann0see** [comment]
Still needs testing on macOS.

**Comment by ann0see** [comment]
Ok. Client and server run on macOS. But https://github.com/jamulussoftware/jamulus/issues/3287 shows up and may be even worse: I cannot register to a directory on macOS at all!

**Comment by ann0see** [comment]
@pljones I believe this is only a minor change (even though the build architecture changed). It runs on macOS non legacy Intel as expected. The build times may decrease.

File: .github/autobuild/mac.sh [hunk 3]
```diff
@@ -47,7 +47,15 @@ setup() {
         echo "Using Qt installation from previous run (actions/cache)"
     else
         echo "Installing Qt..."
-        python3 -m pip install "aqtinstall==${AQTINSTALL_VERSION}"
+        # We may need to create the Qt installation directory and chown it to the runner user to fix permissions
+        sudo mkdir "${QT_DIR}"
```
Comment (pljones) [inline .github/autobuild/mac.sh:51]:
```suggestion
        sudo mkdir -p "${QT_DIR}"
```

**Review by pljones — APPROVED** [review]

**Review by pljones — APPROVED** [review]

**Review by softins — APPROVED** [review]
Changes look reasonable and the CI build succeeds.

**Comment by softins** [comment]
Just wanted to add a point of interest. I just tried rebasing and running a test build of mine that runs CodeQL on the main Mac build instead of on the legacy build. It would appear that CodeQL can't be enabled on the ARM runner. At least maybe not without other changes. See https://github.com/softins/jamulus/actions/runs/9874046539

**Comment by ann0see** [comment]
I think it's not ARM but rather code signing. We moved codeql to legacy once in the past.

**Comment by softins** [comment]
> I think it's not ARM but rather code signing. We moved codeql to legacy once in the past.

CodeQL for Mac currently _is_ on legacy, and has been for a long time. I was just experimenting with moving it back to the main build, which worked after the `sudo` change to the signing code, and prior to the change to macos-14 on ARM.

The error messages on the run I linked to above suggest that the problem is with the ARM building environment, not the signing. Something to do with Rosetta.

But it's not a big problem to leave it as it is; I just thought it was good for us to be aware of the limitation.

**Pull request merged**